### PR TITLE
Source Engine Query Cacher v2.0.0 Release

### PR DESCRIPTION
**Pull request by Michael-Phoenix**
- Added A2S_INFO Challenge/Response
- Added A2S_RULES Challenge/Response
- Fixed Inconsistencies in config/command line processing
- Fixed basic tests for all cases (could be a lot more, especially negative testing)
- replaced hard-coded packet lengths and offsets with constants from PACKETS class
- changed A2S_PLAYERS to use generic A2S_CHALLENGE Code
- complete CI/CD cycle with MAVEN and surefire in GitLab
- this is just a reupload to github to make the PR possible, so there might be changes needed to readme etc after the Pull (check comments)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,15 +1,21 @@
 # Source Engine Query Cacher
-![Downloads](
-https://img.shields.io/github/downloads/hyperxpro/SourceEngineQueryCacher/total)
 
-<p> High-Performance Source Engine Query Cacher for caching and responding A2S_INFO and A2S_PLAYER packets. </p>
+## This is a hard fork of
```
Comment (Michael-Phoenix) [inline README.md:3]: This probably has to be reworked if you decide to approve the PR. I wasn't certain if you were interested so I made sure people get referred to you if the stumble across this on other platforms like GitLab

File: Cacher.conf [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
-SendBufferSize=65535
+Threads=1
+StatsPPS=true
+StatsbPS=true
+GameUpdateRate=1000
+GameUpdateTimeout=1000
+ChallengeCodeTTL=5000
+LocalServerIPAddress=127.0.0.1
+LocalServerPort=43001
+GameServerIPAddress=88.198.1.79
```
Comment (Michael-Phoenix) [inline Cacher.conf:9]: GameServerIPAddress sould probably be reset to 127.0.0.1 or whatever you use in your tests

File: README.md [hunk 3]
```diff
@@ -1,7 +1,6 @@
 # Source Engine Query Cacher
 
-## This is a hard fork of
-https://github.com/hyperxpro/SourceEngineQueryCacher
+## This is a hard fork of https://github.com/hyperxpro/SourceEngineQueryCacher
```
Comment (Michael-Phoenix) [inline README.md:3]: This probably has to be reworked if you decide to approve the PR. I wasn't certain if you were interested so I made sure people get referred to you if they stumble across this on other platforms like GitLab

**Comment by hyperxpro** [comment]
Thanks a lot for the work!

I'm currently on vacation so I'm unable to review it right now. But from a quick view, it looks like this will be SEQC v2.0 because of the number of changes.

**Comment by ThePureCC** [comment]
@Michael-Phoenix Thank you for the PR. As a user I really appreciate your effort. 

Is it possible for you to release a binary of the jar file for me to run tests on it?

**Comment by Michael-Phoenix** [comment]
> @Michael-Phoenix Thank you for the PR. As a user I really appreciate your effort.
> 
> Is it possible for you to release a binary of the jar file for me to run tests on it?

Will upload a jar tomorrow morning. I’m running it on a live 100 player SquadGame Server, which was recently target of lots of attacks and so far the cacher is a godsend.

**Comment by Michael-Phoenix** [comment]
@pureche Here's the lates package that comes out of GitLab's CI/CD:

https://gitlab.com/Michael-Phoenix/bb_sourceenginequerycacher/-/package_files/20774144/download

And here's the test and build log:
https://gitlab.com/Michael-Phoenix/bb_sourceenginequerycacher/-/jobs/1756982544

This build still has the problem that the server sometimes can't be seen on the browser.

**Comment by Michael-Phoenix** [comment]
This one fixes servers not reliably showing up on client server browsers:
https://github.com/Michael-Phoenix/SourceEngineQueryCacher/releases/tag/1.7.1

**Comment by ThePureCC** [comment]
@Michael-Phoenix Thanks. I'll test it out later. Been really busy with stuff.

**Comment by hyperxpro** [comment]
The code looks good so far, I will make some changes before merging and releasing SEQC v2. In meantime, please test this and let us know how well is this performing.

**Comment by Michael-Phoenix** [comment]
@hyperxpro 
I've worked on the program some more over the last week.
We've also commenced some stress tests and figured some more stuff out that probably needs to be implemented.
Shoud I cancel this PR and do a new one?

**Comment by hyperxpro** [comment]
@Michael-Phoenix Sure thing, go ahead. :)

Do let me know if you've any question.

**Comment by Michael-Phoenix** [comment]
@hyperxpro We ran attacks of up to 900k packets per second towards the server this weekend. For a stress test.
When the packets get dropped at the iptables level, the machine could handle up to 900k pps. But when we went and sent proper spoofed A2S requests, we ran into issues like huge memory footprint for the challenge code cacher, even though the TTL was only 2 seconds. I'm currently trying to lower the memory footprint of the Mapped challenge codes and corresponding IPs. IPs will be stored as byte[4] (ipv4) now and I could indeed use some of your experience. You don't happen to have a discord account? Would love to hop in a voip with you and bounce some ideas.

**Comment by hyperxpro** [comment]
@Michael-Phoenix Sure, let's hop on Discord. #HyperX_Pro#4804

**Comment by ThePureCC** [comment]
@Michael-Phoenix 

Switching from 1.6.6, issue is with 1.7.1

image


image

Not getting any packets. Checked iptables, the traffic was from 1.6.6. Server also not showing up in server list. 

Probably something stupid that I did, but I don't know what it is.

**Comment by hyperxpro** [comment]
@pureche 

From a quick look, it looks like you've not enabled `localhost` forwarding in your server.
Try this: `sysctl -w net.ipv4.conf.all.route_localnet=1`

**Comment by ThePureCC** [comment]
> @pureche
> 
> From a quick look, it looks like you've not enabled `localhost` forwarding in your server. Try this: `sysctl -w net.ipv4.conf.all.route_localnet=1`

@hyperxpro 
I ran it, still same result.

Edit:

Unless the dependencies had changed?

**Comment by Michael-Phoenix** [comment]
@hyperxpro I'm getting an error in testing with the recent changes. 
```
16020 [ERROR] A2SPlayerChallenge  Time elapsed: 0.021 s  <<< ERROR!
java.lang.StringIndexOutOfBoundsException: begin 0, end 10, length 8
	at com.aayushatharva.sourcecenginequerycacher.HandlerTest.A2SPlayerChallenge(HandlerTest.java:99)
16023 [ERROR] A2SRulesChallenge  Time elapsed: 0.006 s  <<< ERROR!
java.lang.StringIndexOutOfBoundsException: begin 0, end 10, length 8
	at com.aayushatharva.sourcecenginequerycacher.HandlerTest.A2SRulesChallenge(HandlerTest.java:141)
16357 [INFO] 
16358 [INFO] Results:
16358 [INFO] 
16358 [ERROR] Errors: 
16359 [ERROR]   HandlerTest.A2SPlayerChallenge:99 » StringIndexOutOfBounds begin 0, end 10, le...
16359 [ERROR]   HandlerTest.A2SRulesChallenge:141 » StringIndexOutOfBounds begin 0, end 10, le...
16360 [INFO] 
16361 [ERROR] Tests run: 6, Failures: 0, Errors: 2, Skipped: 0
```
Any idea what might cause the packets sent to only get returned as length of 8?
I've checked and added some debug outputs. The packets seem to get sent in full length.

**Comment by hyperxpro** [comment]
Tests need to be fixed actually.

File: src/main/java/com/aayushatharva/sourcecenginequerycacher/gameserver/a2splayer/PlayerClient.java [hunk 4]
```diff
@@ -31,15 +31,14 @@ public void run() {
                     .option(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
                     .option(ChannelOption.SO_SNDBUF, Config.SendBufferSize)
                     .option(ChannelOption.SO_RCVBUF, Config.ReceiveBufferSize)
-                    .option(EpollChannelOption.MAX_DATAGRAM_PAYLOAD_SIZE, 1400)
```
Comment (Michael-Phoenix) [inline src/main/java/com/aayushatharva/sourcecenginequerycacher/gameserver/a2splayer/PlayerClient.java]: We need some sort of limiter in packet size here. 
What happens with a full 100player server is that packet size grows beyond the 1400bytes that valve proposed in the protocol. 

and if you’re on the same machine the cacher can actually receive a say 1580byte udp datagram internally. 
But as soon as you try and send that over the internet it gets fragmented. And fragmented UDP packets get dropped at router level in 99% of today’s routers. 


so we either implement a proper multi packet handling as it is proposed in the protocol, or we truncate the packet down to 1400 bytes

And just to make sure We should do that for all three packet types.

**Review by Michael-Phoenix — COMMENTED** [review]
This needs to be addressed before being pushed.

File: src/main/java/com/aayushatharva/sourcecenginequerycacher/gameserver/a2splayer/PlayerClient.java (hunk 4 above)
Comment (hyperxpro) [reply src/main/java/com/aayushatharva/sourcecenginequerycacher/gameserver/a2splayer/PlayerClient.java]: Alright then. Let's restore it.

**Comment by Michael-Phoenix** [comment]
> Tests need to be fixed actually.

Did you manage to fix the broken tests?

**Comment by hyperxpro** [comment]
> > Tests need to be fixed actually.
> 
> Did you manage to fix the broken tests?

Unfortunately no. Did you?

**Comment by Michael-Phoenix** [comment]
> 

I‘ve spent three hours on it on saturday but I don‘t even know what changed the behavior in the first place. 
And I was expecting you ran the unit tests before pushing to master. 
Do you have an idea which of your changes might have altered it?

**Comment by hyperxpro** [comment]
> > 
> 
> I‘ve spent three hours on it on saturday but I don‘t even know what changed the behavior in the first place. And I was expecting you ran the unit tests before pushing to master. Do you have an idea which of your changes might have altered it?

I'm on it.

**Comment by xSL0W** [comment]
Hi, if someone has a working version with a2s challenges please let me know.

**Comment by hyperxpro** [comment]
@xSL0W Are you interested in beta testing of the new A2S INFO Challenge?

**Comment by xSL0W** [comment]
> @xSL0W Are you interested in beta testing of the new A2S INFO Challenge?

Yes, DM me on discord when you have a good version for testing
xSLOW#0508

**Comment by toorisrael** [comment]
@hyperxpro I'm also interested in testing this, hopefully with all options working from commandline again and option to disable A2S_INFO challange. I've tried building multiple branches by myself, but I'm a total noob in Java (actually this is my first experience with Java). All versions seems to have some bugs, v1.6.6 seems to work quite well for me, but has one bug:
image
It shows wrong score / time for one, last player (1/2 queries are wrong - packet is cut before the end). I guess it's fixed in 2.0.0 looking at code changes, but this version (v2-wip) doesn't work for me at all, and is also missing commandline options.

**Comment by hyperxpro** [comment]
@toorisrael Hit me on Discord: HyperX_Pro#4804

V2 is not ready that's why you can't compile and use it yet. I'll share the beta release on Discord once ready.

**Pull request merged**